src/Utils.cc

```diff
@@ -563,4 +574,271 @@ thread_type_t fbgemmGetThreadPartition(
+      K key = input_keys[i];
+      local_histogram[(key >> (pass * 8)) & 0xFF]++;
+    }
+  }
+#pragma omp barrier
+  // Step 2: prefix sum
+  if (tid == 0) {
+    if (pass_with_sign_bit) {
+      combine_prefix_sum_for_msb(
+          nthreads, elements_count, histogram, histogram_ps);
+
+    } else {
+      combine_prefix_sum(nthreads, elements_count, histogram, histogram_ps);
+    }
+  }
+#pragma omp barrier
+
+  // Step 3: scatter
+#pragma omp for schedule(static)
+  for (int64_t i = 0; i < elements_count_4; i += 4) {
+    K key_1 = input_keys[i];
+    K key_2 = input_keys[i + 1];
+    K key_3 = input_keys[i + 2];
+    K key_4 = input_keys[i + 3];
+
+    const int bin_1 = (key_1 >> (pass * 8)) & 0xFF;
+    const int bin_2 = (key_2 >> (pass * 8)) & 0xFF;
+    const int bin_3 = (key_3 >> (pass * 8)) & 0xFF;
+    const int bin_4 = (key_4 >> (pass * 8)) & 0xFF;
+
+    int pos;
+    pos = local_histogram_ps[bin_1]++;
+    output_keys[pos] = key_1;
+    output_values[pos] = input_values[i];
+    pos = local_histogram_ps[bin_2]++;
+    output_keys[pos] = key_2;
+    output_values[pos] = input_values[i + 1];
+    pos = local_histogram_ps[bin_3]++;
+    output_keys[pos] = key_3;
+    output_values[pos] = input_values[i + 2];
+    pos = local_histogram_ps[bin_4]++;
+    output_keys[pos] = key_4;
+    output_values[pos] = input_values[i + 3];
+  }
+  if (tid == (nthreads - 1)) {
+    for (int64_t i = elements_count_4; i < elements_count; ++i) {
+      K key = input_keys[i];
+      const int pos = local_histogram_ps[(key >> (pass * 8)) & 0xFF]++;
+      output_keys[pos] = key;
+      output_values[pos] = input_values[i];
+    }
+  }
+}
+
+} // namespace
+
+template <typename K, typename V>
+std::pair<K*, V*> radix_sort_parallel(
+    K* inp_key_buf,
+    V* inp_value_buf,
+    K* tmp_key_buf,
+    V* tmp_value_buf,
+    int64_t elements_count,
+    int64_t max_value,
+    bool maybe_with_neg_vals) {
+  if (max_value == 0) {
+    return std::make_pair(inp_key_buf, inp_value_buf);
+  }
+
+  const int maxthreads = omp_get_max_threads();
+#ifdef _MSC_VER
+  const size_t array_size = (size_t)RDX_HIST_SIZE * maxthreads;
+  // fixes MSVC error C2131
+  int* histogram = static_cast<int*>(
+      fbgemm::fbgemmAlignedAlloc(64, (array_size) * sizeof(int)));
+  int* histogram_ps = static_cast<int*>(
+      fbgemm::fbgemmAlignedAlloc(64, (array_size + 1) * sizeof(int)));
+
+#else
+  alignas(64) int histogram[RDX_HIST_SIZE * maxthreads];
+  alignas(64) int histogram_ps[RDX_HIST_SIZE * maxthreads + 1];
+#endif
+  // If negative values are present, we want to perform all passes
+  // up to a sign bit
+  int num_bits = sizeof(K) * 8;
+  if (!maybe_with_neg_vals)
+    // __builtin_clz is not portable
+    num_bits -= count_leading_zeros(
+        static_cast<typename std::make_unsigned<K>::type>(max_value));
+
+  const unsigned int num_passes = (num_bits + 7) / 8;
+
+#pragma omp parallel
+  {
+    K* input_keys = inp_key_buf;
+    V* input_values = inp_value_buf;
```

Comment:
```suggestion
    const K* input_keys = inp_key_buf;
    const V* input_values = inp_value_buf;
```

Reply:
I believe it is incorrect, for the same reasons as described [here](https://github.com/pytorch/FBGEMM/pull/1672#discussion_r1160496180).